src/shared_utilities/interpolating_time_varying_input1d.jl

```diff
@@ -0,0 +1,138 @@
+
+"""
+    LinearInterpolation
+
+Perform linear interpolation between the two neighboring points.
+"""
+struct LinearInterpolation <: AbstractInterpolationMethod end
+
+"""
+    InterpolatingTimeVaryingInput1D
+
+The constructor for InterpolatingTimeVaryingInput1D is not supposed to be used directly, unless you
+know what yo are doing. The constructor does not perform any check and does not take care of
+GPU compatibility. It is repsonsability of Integrator() to do so.
+
+"""
+struct InterpolatingTimeVaryingInput1D{
+    AA1 <: AbstractArray,
+    AA2 <: AbstractArray,
+    M <: AbstractInterpolationMethod,
+    CC <: ClimaComms.AbstractCommsContext,
+    R <: Tuple,
+} <: AbstractTimeVaryingInput
+    # AA1 and AA2 could be different because of different FTs
+
+    """Independent coordinate"""
+    times::AA1
+
+    """Variable"""
+    vals::AA2
+
+    """Interpolation method"""
+    method::M
+
+    """ClimaComms context"""
+    context::CC
+
+    """Range of times over which the interpolator is defined. range is always defined on the
+    CPU. Used by the in() function."""
+    range::R
+end
+
+# GPU compatibility
+function Adapt.adapt_structure(to, itp::InterpolatingTimeVaryingInput1D)
+    times = Adapt.adapt_structure(to, itp.times)
+    vals = Adapt.adapt_structure(to, itp.vals)
+    method = Adapt.adapt_structure(to, itp.method)
+    range = Adapt.adapt_structure(to, itp.range)
+    # On a GPU, we have a "ClimaCore.DeviceSideContext"
+    InterpolatingTimeVaryingInput1D(
+        times,
+        vals,
+        method,
+        DeviceSideContext(),
+        range,
+    )
+end
+
+function evaluate!(destination, itp::InterpolatingTimeVaryingInput1D, time)
+    time in itp || error("TimeVaryingInput does not cover time $time")
+    if ClimaComms.device(itp.context) isa ClimaComms.CUDADevice
+        CUDA.@cuda evaluate!(parent(destination), itp, time, itp.method)
+    else
+        evaluate!(parent(destination), itp, time, itp.method)
+    end
+    return nothing
+end
+
+function TimeVaryingInput(
+    times::AbstractArray,
+    vals::AbstractArray;
+    method = LinearInterpolation(),
+    context = ClimaComms.context(),
+)
+    issorted(times) || error("Can only interpolate with sorted times")
+    length(times) == length(vals) ||
+        error("times and vals have different lengths")
+
+    # When device is CUDADevice, ArrayType will be a CUDADevice, so that times and vals get
+    # copied to the GPU.
+    ArrayType = ClimaComms.array_type(ClimaComms.device(context))
+
+    range = (times[begin], times[end])
+    return InterpolatingTimeVaryingInput1D(
+        ArrayType(times),
+        ArrayType(vals),
+        method,
+        context,
+        range,
+    )
+end
+
+"""
+    in(time, itp::InterpolatingTimeVaryingInput1D)
+
+Check if the given `time` is in the range of definition for `itp`.
+"""
+function Base.in(time, itp::InterpolatingTimeVaryingInput1D)
+    return itp.range[1] <= time <= itp.range[2]
+end
+
+function evaluate!(
```

Comment:
I added a few lines. I also realized that I don't need one of the edge cases conditions that I added because I check that time is in the range of validity.

Reply:
Actually, I did miss an edge case.